### PR TITLE
Squashed changes for AIRBIN target

### PR DESCRIPTION
**Pull request by jgmelber**
Creates a new target to lower mlir-aie code to a fat binary format, this is supported in mlir-air by runtime functions, test and demonstration cases. The target is derived from `AIRTargetLibXAIEv1.cpp`. The pass is enabled in `aiecc.py` with the flag `--aie-generate-airbin`

### CODE REVIEW

File: lib/Targets/AIETargetAirbin.cpp [hunk 1]
```diff
@@ -0,0 +1,1746 @@
...
+  bool operator<(const Write &rhs) const {
+    return static_cast<uint64_t>(address) < static_cast<uint64_t>(rhs.address);
+  }
+
+private:
+  Address address;
+  uint32_t data;
+};
+
+static std::vector<Write> writes;
```
Comment (stephenneuendorffer) [inline lib/Targets/AIETargetAirbin.cpp:154]: Seems like all this static stuff should be a class, hiding the implementation of 'writes'?

Comment (stephenneuendorffer) [reply lib/Targets/AIETargetAirbin.cpp:154]: At the very least, it would seem that sorting this by write address would be a good thing?

File: lib/Targets/AIETargetAirbin.cpp [hunk 2]
```diff
@@ -0,0 +1,1746 @@
...
+  }
+}
+
+static constexpr auto BD_BASE = 0x1D000u;
+static constexpr auto BD_END = 0x1D1F8u;
+static constexpr auto SHIM_BD_END = 0x1D15Cu;
+
+// The SHIM row is always 0.
+// SHIM resets are handled by the runtime.
+static void generateShimConfig(TileOp &tileOp) {
```
Comment (stephenneuendorffer) [inline lib/Targets/AIETargetAirbin.cpp:233]: assert that it is a shim tile?

File: lib/Targets/AIETargetAirbin.cpp [hunk 3]
```diff
@@ -0,0 +1,1746 @@
...
+  static constexpr auto shifted_mask = unshifted_mask << low_bit;
+
+  [[nodiscard]] constexpr uint32_t operator()(uint32_t value) const {
+    return (value << low_bit) & shifted_mask;
+  }
+};
+
+// LLVM may have a way to read ELF files,
+// but for our purposes, manually parsing is fine for now.
+static bool loadElf(TileAddress tile, const std::string &filename) {
```
Comment (stephenneuendorffer) [inline lib/Targets/AIETargetAirbin.cpp:272]: Um... yeah... There's enough magic stuff here to use the LLVM way... you want:

  Expected<std::unique_ptr<ObjectFile>> ELFObjOrErr =
      object::ObjectFile::createELFObjectFile(
          MemoryBufferRef(toStringRef(D.Data), "dummyELF"));

See llvm/object/ObjectFile.h and llvm/object/ELF.h

File: lib/Targets/AIETargetAirbin.cpp [hunk 4]
```diff
@@ -0,0 +1,1746 @@
...
+      generateNonShimConfig(tileOp);
+    }
+  }
+}
+
+// Start execution of all the cores.
+// This is by the runtime. We do not need to record it.
+/*
+static void start_cores(mlir::ModuleOp module) {
+
```
Comment (stephenneuendorffer) [inline lib/Targets/AIETargetAirbin.cpp:436]: Remove?

File: lib/Targets/AIETargetAirbin.cpp [hunk 5]
```diff
@@ -0,0 +1,1746 @@
...
+static constexpr auto TILE_ADDR_ROW_SHIFT = TILE_ADDR_OFF_WITDH;
+static constexpr auto TILE_ADDR_ROW_WIDTH = 5u;
+
+static constexpr auto TILE_ADDR_COL_SHIFT =
+    TILE_ADDR_ROW_SHIFT + TILE_ADDR_ROW_WIDTH;
+static constexpr auto TILE_ADDR_COL_WIDTH = 7u;
+
+static constexpr auto TILE_ADDR_ARR_SHIFT =
+    TILE_ADDR_COL_SHIFT + TILE_ADDR_COL_WIDTH;
+
```
Comment (stephenneuendorffer) [inline lib/Targets/AIETargetAirbin.cpp:66]: Perhaps this magic stuff should come out of libXAIE?

File: lib/Targets/AIETargetAirbin.cpp [hunk 6]
```diff
@@ -0,0 +1,1746 @@
...
+  };
+
+  Field<1> dmaChannelReset;
+  Field<0> dmaChannelEnable;
+
+  for (auto memOp : module.getOps<MemOp>()) {
+    /* clang-format off
+    output << "XAieDma_TileInitialize(" << tileInstStr(col, row) << ", " << tileDMAInstStr(col, row) << ");\n";
+    ----
+    DmaInstPtr->BaseAddress = TileInstPtr->TileAddr;
```
Comment (stephenneuendorffer) [inline lib/Targets/AIETargetAirbin.cpp:523]: Remove this commented out code?

File: tools/aiecc/aiecc/main.py [hunk 7]
```diff
@@ -320,10 +324,12 @@ async def run_flow(self):
         progress.update(progress.task,advance=0,visible=False)
         progress.task_completed = progress.add_task("[green] AIE Compilation:", total=len(cores)+1, command="%d Workers" % nworkers)
 
-        processes = [self.process_arm_cgen()]
+        self.progress_bar = progress
+        processes = []
         for core in cores:
           processes.append(self.process_core(core))
         await asyncio.gather(*processes)
+        await self.process_arm_cgen()
```
Comment (stephenneuendorffer) [inline tools/aiecc/aiecc/main.py:333]: Why this change?  The intention here is to start running ARM code generation ASAP, since it is slower than everything else.  If we can't do that here because the generated ELF files need to be parsed, then we should probably generate a header and separately compile and link the host code in 2 parts.

File: tools/aiecc/aiecc/main.py [hunk 8]
```diff
@@ -221,11 +221,15 @@ async def process_arm_cgen(self):
...
-      file_inc_cpp = os.path.join(self.tmpdirname, 'aie_inc.cpp')
-      if(opts.xaie == 1):
-          await self.do_call(task, ['aie-translate', '--aie-generate-xaie', '--xaie-target=v1', file_physical, '-o', file_inc_cpp])
+      if(opts.airbin):
+        file_airbin = os.path.join(self.tmpdirname, 'air.bin')
+        await self.do_call(task, ['aie-translate', '--aie-generate-airbin', file_physical, '-o', file_airbin])
       else:
-          await self.do_call(task, ['aie-translate', '--aie-generate-xaie', '--xaie-target=v2', file_physical, '-o', file_inc_cpp])
+        file_inc_cpp = os.path.join(self.tmpdirname, 'aie_inc.cpp')
+        if(opts.xaie == 1):
```
Comment (stephenneuendorffer) [inline tools/aiecc/aiecc/main.py:229]: As I read this, you're no longer generating aie_inc.cpp... But there's a bunch of host APIs in there we probably can't just drop.

**Review by stephenneuendorffer — CHANGES_REQUESTED** [review]
Overall, I'm a little surprised by the result here..  I was expecting something that still depended on libXAIE... instead it seems that the modified libXAIE backend was used to reverse engineer what it was doing, and the logic has been replicated here.... This seems like a significant piece of code that will need to be maintained indefinitely and also will change for AIEv2, etc.

File: tools/aiecc/aiecc/main.py (hunk 7 above)
Comment (jgmelber) [reply tools/aiecc/aiecc/main.py:333]: The change here is that the airbin generation occurs in `self.process_arm_cgen()` and it needs to have the AIE elfs already in order to generate the airbin

File: tools/aiecc/aiecc/main.py (hunk 8 above)
Comment (jgmelber) [reply tools/aiecc/aiecc/main.py:229]: For AIR we don't need those APIs when we are using airbin, air_inc.cpp is only used for configuration via libXAIE

Comment (stephenneuendorffer) [reply tools/aiecc/aiecc/main.py:229]: So all the tests run with airbin enabled?  There are more host APIs generated than just configuration.

File: lib/Targets/AIETargetAirbin.cpp (hunk 5 above)
Comment (jgmelber) [reply lib/Targets/AIETargetAirbin.cpp:66]: Agreed, it would be best if the architecture specific values were factored out of all of the AIETargets

File: lib/Targets/AIETargetAirbin.cpp (hunk 1 above)
Comment (jgmelber) [reply lib/Targets/AIETargetAirbin.cpp:154]: The writes are sorted at the end before generating the binary

File: tools/aiecc/aiecc/main.py (hunk 8 above)
Comment (jgmelber) [reply tools/aiecc/aiecc/main.py:229]: Which tests? This works for the mlir-air tests using airbin, it has not been tested on every air test/example yet. It is not intended as a back end for mlir-aie tests without the air runtime

File: tools/aiecc/aiecc/main.py (hunk 7 above)
Comment (stephenneuendorffer) [reply tools/aiecc/aiecc/main.py:333]: I'd prefer to see this separated out.

File: tools/aiecc/aiecc/main.py (hunk 8 above)
Comment (stephenneuendorffer) [reply tools/aiecc/aiecc/main.py:229]: But seems like it very easily could be...

**Comment by keryell** [comment]
Status of this PR?
